### ISSUE TITLE
get_cached_content fails with 'SafeString object has no attribute append' error

### ISSUE THREAD

**Issue by jenda1** [django-wiki/django-wiki#1072]
`get_cached_content()` expects the `cached_items` is a list. It is... often. But sometimes it is a SafeString and then the line https://github.com/django-wiki/django-wiki/blob/e5d6b9701f9cef4323de7748a037f78f323c11a8/src/wiki/models/article.py#L258 fails because string does not have append method.

It looks that the cache content is set outside the `get_cached_content` method sometimes, but I do not know where...

There is the exception:

```
  File "/tmp/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/tmp/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 145, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/tmp/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 143, in _get_response
    response = response.render()
  File "/tmp/venv/lib/python3.8/site-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File "/tmp/venv/lib/python3.8/site-packages/django/template/response.py", line 83, in rendered_content
    return template.render(context, self._request)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 171, in render
    return self._render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/test/utils.py", line 95, in instrumented_test_render
    return self.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/test/utils.py", line 95, in instrumented_test_render
    return self.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/test/utils.py", line 95, in instrumented_test_render
    return self.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/test/utils.py", line 95, in instrumented_test_render
    return self.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/classytags/core.py", line 151, in render
    return self.render_tag(context, **kwargs)
  File "/tmp/venv/lib/python3.8/site-packages/sekizai/templatetags/sekizai_tags.py", line 87, in render_tag
    rendered_contents = nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 936, in render
    bit = node.render_annotated(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/base.py", line 903, in render_annotated
    return self.render(context)
  File "/tmp/venv/lib/python3.8/site-packages/django/template/library.py", line 214, in render
    _dict = self.func(*resolved_args, **resolved_kwargs)
  File "/tmp/venv/lib/python3.8/site-packages/wiki/templatetags/wiki_tags.py", line 53, in wiki_render
    content = article.get_cached_content(user=context.get("user"))
  File "/tmp/venv/lib/python3.8/site-packages/wiki/models/article.py", line 258, in get_cached_content
    cached_items.append(cache_content_key)
AttributeError: 'SafeString' object has no attribute 'append'
```

**Comment by benjaoming** [comment]
Thanks for reporting @jenda1 - seems like an issue that could perhaps be addressed with a test case and a fix.

**Comment by Steckelfisch** [comment]
me too:
`Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/

Django Version: 2.2.16
Python Version: 3.8.5
Installed Applications:
['django.contrib.humanize.apps.HumanizeConfig',
 'django.contrib.auth.apps.AuthConfig',
 'django.contrib.contenttypes.apps.ContentTypesConfig',
 'django.contrib.sessions.apps.SessionsConfig',
 'django.contrib.sites.apps.SitesConfig',
 'django.contrib.messages.apps.MessagesConfig',
 'django.contrib.staticfiles.apps.StaticFilesConfig',
 'django.contrib.admin.apps.AdminConfig',
 'django.contrib.admindocs.apps.AdminDocsConfig',
 'sekizai',
 'sorl.thumbnail',
 'django_nyt.apps.DjangoNytConfig',
 'wiki.apps.WikiConfig',
 'wiki.plugins.macros.apps.MacrosConfig',
 'wiki.plugins.help.apps.HelpConfig',
 'wiki.plugins.links.apps.LinksConfig',
 'wiki.plugins.images.apps.ImagesConfig',
 'wiki.plugins.attachments.apps.AttachmentsConfig',
 'wiki.plugins.notifications.apps.NotificationsConfig',
 'wiki.plugins.editsection.apps.EditSectionConfig',
 'wiki.plugins.globalhistory.apps.GlobalHistoryConfig',
 'mptt']
Installed Middleware:
['django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware']


Template error:
In template /home/braas/projects/django_wiki/django-wiki/src/wiki/templates/wiki/base_site.html, error at line 0
   'SafeText' object has no attribute 'append'
   1 : {% load sekizai_tags i18n wiki_tags static %}<!DOCTYPE html>
   2 : <html lang="en">
   3 :   <head>
   4 :     <meta charset="utf-8">
   5 :     <title>{% block wiki_pagetitle %}{% endblock %}{% block wiki_site_title %} - django-\/\/  i K |{% endblock %}</title>
   6 :     <meta name="viewport" content="width=device-width, initial-scale=1.0">
   7 :     <meta http-equiv="X-UA-Compatible" content="IE=edge">
   8 :     <meta name="description" content="">
   9 :     <meta name="author" content="www.django-wiki.org">
   10 : 


Traceback:

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/core/handlers/base.py" in _get_response
  145.                 response = self.process_exception_by_middleware(e, request)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/core/handlers/base.py" in _get_response
  143.                 response = response.render()

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/response.py" in render
  106.             self.content = self.rendered_content

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/response.py" in rendered_content
  83.         content = template.render(context, self._request)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/backends/django.py" in render
  61.             return self.template.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  171.                     return self._render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/classytags/core.py" in render
  151.         return self.render_tag(context, **kwargs)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/sekizai/templatetags/sekizai_tags.py" in render_tag
  87.         rendered_contents = nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/braas/projects/django_wiki/ve-djang-wiki/lib/python3.8/site-packages/django/template/library.py" in render
  214.         _dict = self.func(*resolved_args, **resolved_kwargs)

File "/home/braas/projects/django_wiki/django-wiki/src/wiki/templatetags/wiki_tags.py" in wiki_render
  53.         content = article.get_cached_content(user=context.get("user"))

File "/home/braas/projects/django_wiki/django-wiki/src/wiki/models/article.py" in get_cached_content
  262.         cached_items.append(cache_content_key)

Exception Type: AttributeError at /
Exception Value: 'SafeText' object has no attribute 'append'`

**Comment by jenda1** [comment]
If there is agreement, I'll remove the "per-article-user" caching schema. It simplify the code, speed up the wiki, and fix this issue.  I'll revert to "per-article" caching.

TLDR; the "per-article-user" caching schema was added in https://github.com/django-wiki/django-wiki/commit/8cc19e5bc7814b622e267b1a56234a566e33cde2 by me. The idea was to allow the creation of a plugin that renders the page differently for each user. But I have not created such plugin, I came up with an easier way to achieve what I needed.  AFIK there is no plugin that uses the functionality.
The "per-article-user" is useless now, it causes problems and it is a bit expensive because cache entry can not be used for another user.

**Comment by Steckelfisch** [comment]
> If there is agreement, I'll remove the "per-article-user" caching schema. It simplify the code, speed up the wiki, and fix this issue. I'll revert to "per-article" caching.

I agree. 
Fyi: for me the problem appeared when i upgraded my codebase to the latest development version

**Comment by iburadempa** [comment]
I use cached sessions with LocMemCache and found the problem to be reproducible. I also couldn't find a cause in the django-wiki code. My guess would be that the cache pollution is related to sessions. But anyway, 1e4b2fd solves the problem and calls for a bugfix release!

**Comment by Steckelfisch** [comment]
I confirm  https://github.com/django-wiki/django-wiki/issues/1072 solves the problem

**Comment by benjaoming** [comment]
#1080 may fix the exception but it will produce wrong cache results because the user session isn't taken into account.

**Comment by Steckelfisch** [comment]
in models/Article.py, near line 260
The first view of an Article shows correctly.
Then, for each Article entry, two entries are stored.
The second view of  the same Article throws the exeption

`cache_key = self.get_cache_key() # value: wikiarticle85nl-nl `
 `cache_content_key = self.get_cache_content_key(user) # value: wikiarticle85nl-nl `
  . . . . .
 `cache.set(cache_key, cached_items, settings.CACHE_TIMEOUT) `
 `cache.set(cache_content_key, cached_content, settings.CACHE_TIMEOUT) `

I think the first cache entry will be overwritten with the second one.

**Comment by Steckelfisch** [comment]
ah. Found it.

the user key is sluggified, and that removes the ':'. AnonymousUser has an empty lastname, so the resulting cache_content_key
is equal to the cache_key.
models/article.py, line 235:
`key_raw = "{key}:{user}".format()`
should be:
`key_raw = "{key}-{user}".format()`

then, everything woks correctly..

**Comment by Steckelfisch** [comment]
I wil revert al the chnages I have made locally, and create a test_cache test in core. Then I will fix the issue and submit a pull request.

**Comment by mesenev** [comment]
My main page gives me 500 because of this. Any workaround for 0.7?

**Comment by benjaoming** [comment]
This issue is back and is confirmed with test cases added recently.
